soc: arm: nxp_lpc: Only clock core from PLL1 when CONFIG_FLASH=n

danieldegrasse opened the pull request:
Do not clock the LPC55xxx cores from PLL1 when CONFIG_FLASH is set. This is required due to the following limitation of the flash controller (documented in the reference manual):

Flash operations (erase, blank check, program) and reading a single word can only be performed for CPU frequencies of up to 100 MHz. These operations cannot be performed for frequencies above 100 MHz.

The PLL1 clock source will result in a core clock of 150MHz, which violates this requirement.

Fixes #62963

danieldegrasse committed:
soc: arm: nxp_lpc: Only clock core from PLL1 when CONFIG_FLASH=n

Do not clock the LPC55xxx cores from PLL1 when CONFIG_FLASH is set. This
is required due to the following limitation of the flash controller
(documented in the reference manual):

Flash operations (erase, blank check, program) and reading a single word
can only be performed for CPU frequencies of up to 100 MHz. These
operations cannot be performed for frequencies above 100 MHz.

The PLL1 clock source will result in a core clock of 150MHz, which
violates this requirement.

Fixes #62963

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

DerekSnell commented:
Hi @danieldegrasse ,
Thank you for proposing this solution.  But I do not like this frequency limitation required on all applications that use the flash driver.  Some apps will only write to the internal flash once during manufacturing.  Others may program the flash very infrequently.

I think a better solution is to update the flash driver for this SOC.  When that driver does an operation, it checks the CPU frequency, and returns an error if the frequency exceeds the limit.  Then let the app deal with the frequency limit.  That allows apps to execute above 100MHz most of the time, and they can dynamically change the frequency if needed before using the flash driver.  Or other apps may choose to always clock at 100MHz or below.  

For the sample and test apps, we can set the Kconfig for this board so `INIT_PLL1=n`.  And add a note why, so developers are aware of the limitation.

But we should allow these options to give the applications the most flexibility.  Thanks

danieldegrasse committed:
boards: arm: lpcxpresso: update core clock documentation

Update documentation for core clock frequency on LPC55xxx boards, to
indicate that they are clocked from PLL1 at 150MHz by default. Add a
note about the reduction to 96MHz required when using the flash
controller.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse commented:
Hi @DerekSnell- I do agree that making users aware they can increase the clock frequency to 150MHz when not using the flash driver makes sense. However, given that the HAL driver for the SOC flash requires the core frequency to be 96MHz at initialization, I think we should likely set this as the default when using the flash subsystem.

I've added some documentation to this PR to help users understand why the clock frequency will be lower when building with `CONFIG_FLASH=y`. Additionally, I will open a ticket internally to add frequency checks to the SDK flash driver, since this issue will affect more than just the users of Zephyr (so we should enforce the restriction there)

DerekSnell commented:
> Hi @DerekSnell- I do agree that making users aware they can increase the clock frequency to 150MHz when not using the flash driver makes sense. However, given that the HAL driver for the SOC flash requires the core frequency to be 96MHz at initialization, I think we should likely set this as the default when using the flash subsystem.
> 
> I've added some documentation to this PR to help users understand why the clock frequency will be lower when building with `CONFIG_FLASH=y`. Additionally, I will open a ticket internally to add frequency checks to the SDK flash driver, since this issue will affect more than just the users of Zephyr (so we should enforce the restriction there)

Thanks @danieldegrasse , I agree with this, and approved the PR.
